extensions/dicom-p10-downloader/src/downloadAndZip.js

```diff
@@ -0,0 +1,190 @@
+ *      be downloaded (instead of all the instances from that series), the list
+ *      could be changed to:
+ *
+ *    [
+ *      'studyUIDFromA',
+ *      ['studyUIDFromB', [
+ *        'seriesUIDFromI',
+ *        ['seriesUIDFromJ', [
+ *          'instanceUIDFromX',
+ *          'instanceUIDFromY',
+ *          'instanceUIDFromZ'
+ *        ]]
+ *      ]]
+ *    ]
+ *
+ * @param {Object} options A plain object with options;
+ * @param {function} options.progress A callback to retrieve notifications
+ * @returns {Promise} A promise that resolves to an URL from which the ZIP file
+ *  can be downloaded;
+ */
+
+async function downloadAndZip(dicomWebClient, listOfUIDs, options) {
+  if (dicomWebClient instanceof api.DICOMwebClient) {
+    const buffers = await downloadAll(dicomWebClient, listOfUIDs, options);
+    return zipAll(buffers, options);
+  }
+  throw new Error('A valid DICOM Web Client instance is expected');
+}
+
+/**
+ * Utils
+ */
+
+async function zipAll(buffers) {
+  const zip = new JSZip();
+  OHIF.log.info('Adding DICOM P10 files to archive:', buffers.length);
+  buffers.forEach((buffer, i) => {
+    const path = buildPath(buffer) || `${i}.dcm`;
+    zip.file(path, buffer);
+  });
+  return URL.createObjectURL(await zip.generateAsync({ type: 'blob' }));
+}
+
+function buildPath(buffer) {
+  let path;
+  try {
+    const byteArray = new Uint8Array(buffer);
+    const dataSet = dicomParser.parseDicom(byteArray, {
+      // Stop parsing after Series Instance UID is found
+      untilTag: 'x0020000e',
+    });
+    const studyUID = dataSet.string('x0020000d');
+    const seriesUID = dataSet.string('x0020000e');
+    const instanceUID = dataSet.string('x00080018');
+    if (studyUID && seriesUID && instanceUID) {
+      path = `${studyUID}/${seriesUID}/${instanceUID}.dcm`;
+    }
+  } catch (e) {
+    OHIF.log.error('Error parsing downloaded DICOM P10 file...', e);
+  }
+  return path;
+}
+
+async function downloadAll(client, listOfUIDs, progress) {
+  const downloads = [];
+
+  // Initiate download for each item of the hierarchical list
+  hierarchicalListUtils.forEach(listOfUIDs, (study, series, instance) => {
+    if (isDicomUid(study)) {
+      downloads.push(download(client, study, series, instance));
+    }
+  });
+
+  // Make sure at least one download was initiated
+  if (downloads.length < 1) {
+    throw new Error('No valid reference to be downloaded');
+  }
+
+  // Print tree of hierarchical references
+  OHIF.log.info('Downloading DICOM P10 files for references:');
+  OHIF.log.info(hierarchicalListUtils.print(listOfUIDs));
+
+  // Wait on created download promises
+  return Promise.all(downloads).then(results => {
+    const buffers = [];
+    // The "results" array may directly contain buffers (ArrayBuffer instances)
+    // or arrays of buffers, depending on the type of downloads initiated on the
+    // previous step (retrieveStudy, retrieveSeries or retrieveinstance). Ex:
```

Comment:
This is making me think we should change dicomweb client later to always return an array for consistency.